File: gee_scripts/03_s2_optical_extraction.js
/*
============================================================
GEE SCRIPT 3 — Sentinel-2 Optical Extraction (Reference + Ground Truth)
SAR-Guided Optical Reconstruction Pipeline (merged)
============================================================
For every patch in both grids, pulls TWO separate optical scenes:

  1. REFERENCE (input) — the most recent cloud-free Sentinel-2 scene
     BEFORE each gap window starts. 6 optical bands + 1 cloud-mask band
     (7 bands total) — the cloud mask here describes clouds in the
     REFERENCE scene itself (used as an input feature in Python, telling
     the model where its reference is unreliable), NOT a synthetic
     blackout.

  2. GROUND TRUTH (target) — a genuinely separate cloud-free Sentinel-2
     scene that exists for a date INSIDE or very near the gap window.
     These are RARE by definition (that's the whole reason this project
     exists) — most patches will have NO ground truth for most windows.
     That's expected and fine; module1_build_manifest.py explicitly
     tracks has_gt=True/False and only has_gt=True rows enter your
     reported quantitative metrics.

Filenames:
  s2ref_<patch_id>_<window_start>.tif   (7 bands)
  s2gt_<patch_id>_<window_start>.tif    (6 bands, only written if found)

These MUST land in folders named exactly optical_reference/ and
optical_ground_truth/ after downloading, matching
module1_build_manifest.py's expected structure.

ACTION REQUIRED: GAP_WINDOWS must exactly match script 2 and
module1_build_manifest.py.
*/

var GEE_USERNAME = 'your_gee_username';  // TODO: SET THIS — must match scripts 1 & 2

var GAP_WINDOWS = [
  ['2021-06-01', '2021-09-30'],
  ['2022-06-01', '2022-09-30'],
  ['2023-06-01', '2023-09-30']
];

var PATCH_SIZE_PX = 256;  // LOCKED
var SCALE_M = 10;

var OPTICAL_BANDS = ['B2', 'B3', 'B4', 'B8', 'B11', 'B12'];
// TODO: confirm this exact order matches training/module2_dataset.py's
// OPTICAL_BAND_ORDER — a mismatch here silently corrupts every NDWI/NDVI
// value computed downstream.

var CLOUD_PROB_THRESHOLD = 30;  // s2cloudless probability % above which a pixel is "cloudy"
var MAX_PRE_GAP_SEARCH_DAYS = 90;  // how far back to search for a clear reference scene

var brahmapurPatches = ee.FeatureCollection('users/' + GEE_USERNAME + '/brahmapur_patch_grid');
var mahanadiPatches = ee.FeatureCollection('users/' + GEE_USERNAME + '/mahanadi_patch_grid');
var allPatches = brahmapurPatches.merge(mahanadiPatches);

// ============================================================
// Cloud masking using s2cloudless
// ============================================================

function getS2WithCloudProb(aoi, startDate, endDate) {
  var s2 = ee.ImageCollection('COPERNICUS/S2_SR_HARMONIZED')
    .filterBounds(aoi)
    .filterDate(startDate, endDate)
    .select(OPTICAL_BANDS);

  var s2Cloud = ee.ImageCollection('COPERNICUS/S2_CLOUD_PROBABILITY')
    .filterBounds(aoi)
    .filterDate(startDate, endDate);

  var joined = ee.Join.saveFirst('cloud_prob').apply({
    primary: s2,
    secondary: s2Cloud,
    condition: ee.Filter.equals({leftField: 'system:index', rightField: 'system:index'})
  });

  return ee.ImageCollection(joined).map(function(img) {
    var cloudProb = ee.Image(img.get('cloud_prob')).select('probability');
    var cloudMask = cloudProb.gte(CLOUD_PROB_THRESHOLD).rename('CLOUD_MASK');
    var cloudFraction = cloudMask.reduceRegion({
      reducer: ee.Reducer.mean(), geometry: aoi, scale: SCALE_M, maxPixels: 1e8
    }).get('CLOUD_MASK');
    return img.addBands(cloudMask).set('cloud_fraction', cloudFraction);
  });
}

function findClearestScene(collection) {
  // Sort by cloud_fraction ascending, take the clearest
  return collection.sort('cloud_fraction').first();
}

// ============================================================
// Per-patch, per-window export loop
// ============================================================

var patchList = allPatches.toList(allPatches.size()).getInfo();
print('Patches to process:', patchList.length);

var refExportCount = 0;
var gtExportCount = 0;
var gtFoundCount = 0;
var gtMissingCount = 0;

patchList.forEach(function(patchFeature) {
  var patchId = patchFeature.properties.patch_id;
  var patchGeom = ee.Geometry(patchFeature.geometry);

  GAP_WINDOWS.forEach(function(window) {
    var windowStart = window[0];
    var windowEnd = window[1];

    // --- 1. Reference scene: clearest in the MAX_PRE_GAP_SEARCH_DAYS before windowStart ---
    var refSearchStart = ee.Date(windowStart).advance(-MAX_PRE_GAP_SEARCH_DAYS, 'day');
    var refCandidates = getS2WithCloudProb(patchGeom, refSearchStart, windowStart);
    var refCount = refCandidates.size().getInfo();

    if (refCount === 0) {
      print('WARNING: no reference scene found for ' + patchId + ' / ' + windowStart + ' — skipping window entirely');
      return;  // no reference -> no usable triplet for this patch/window, matches module1's skip logic
    }

    var refScene = findClearestScene(refCandidates);
    var refImg = ee.Image(refScene).select(OPTICAL_BANDS.concat(['CLOUD_MASK']));

    Export.image.toDrive({
      image: refImg.clip(patchGeom),
      description: 's2ref_' + patchId + '_' + windowStart,
      folder: 'sar_optical_dataset/optical_reference',
      fileNamePrefix: 's2ref_' + patchId + '_' + windowStart,
      region: patchGeom,
      scale: SCALE_M,
      dimensions: PATCH_SIZE_PX + 'x' + PATCH_SIZE_PX,
      crs: 'EPSG:32644',
      maxPixels: 1e9
    });
    refExportCount++;

    // --- 2. Ground truth: rare clear scene INSIDE the gap window itself ---
    var gtCandidates = getS2WithCloudProb(patchGeom, windowStart, windowEnd)
      .filter(ee.Filter.lt('cloud_fraction', 0.05));  // require genuinely clear, <5% cloud
    var gtCount = gtCandidates.size().getInfo();

    if (gtCount === 0) {
      gtMissingCount++;
      return;  // expected and common — this window/patch has no real GT, has_gt=False in manifest
    }

    var gtScene = findClearestScene(gtCandidates);
    var gtImg = ee.Image(gtScene).select(OPTICAL_BANDS);

    Export.image.toDrive({
      image: gtImg.clip(patchGeom),
      description: 's2gt_' + patchId + '_' + windowStart,
      folder: 'sar_optical_dataset/optical_ground_truth',
      fileNamePrefix: 's2gt_' + patchId + '_' + windowStart,
      region: patchGeom,
      scale: SCALE_M,
      dimensions: PATCH_SIZE_PX + 'x' + PATCH_SIZE_PX,
      crs: 'EPSG:32644',
      maxPixels: 1e9
    });
    gtExportCount++;
    gtFoundCount++;
  });
});

print('Reference scene exports queued:', refExportCount);
print('Ground-truth scene exports queued:', gtExportCount);
print('Windows with NO ground truth found (expected, tracked as has_gt=False):', gtMissingCount);
print('Run all tasks from the Tasks tab. After downloading, place files in '
      + 'folders named exactly: optical_reference/ and optical_ground_truth/');

File: gee_scripts/03_s2_optical_extraction.py
"""
GEE SCRIPT 3 (Python) — Sentinel-2 Optical Reference + Ground Truth Extraction
via earthengine-api

Exports two sets of files per patch per gap window:
  1. REFERENCE: clearest S2 scene in the 90 days BEFORE window start
       7 bands: B2, B3, B4, B8, B11, B12, CLOUD_MASK
       filename: s2ref_<patch_id>_<window_start>.tif
       folder:   sar_optical_dataset/optical_reference/

  2. GROUND TRUTH: clearest S2 scene STRICTLY INSIDE the gap window
       with cloud_fraction < 0.05 (sparse by design — most will be skipped)
       6 bands: B2, B3, B4, B8, B11, B12
       filename: s2gt_<patch_id>_<window_start>.tif
       folder:   sar_optical_dataset/optical_ground_truth/

Band order matches module2_dataset.py exactly:
  OPTICAL_BAND_ORDER = ['B2', 'B3', 'B4', 'B8', 'B11', 'B12']
  Reference export: optical bands first (indices 0-5), CLOUD_MASK last (index 6)
  Ground truth export: optical bands only (indices 0-5)

Fixes carried over from Script 2:
  - dimensions only, no scale (avoids GEE Error code 3)
  - CRS = EPSG:32645 (UTM 45N, correct for both AOIs)
  - Chunked .getInfo() calls (CHUNK_SIZE=15) to avoid timeout on
    expensive per-patch cloud-probability reduceRegion calls
"""

import ee
import time

ee.Initialize(project='sar-optical-synthesis')

GEE_USERNAME = 'mohitpradhan'

GAP_WINDOWS = [
    ('2021-06-01', '2021-09-30'),
    ('2022-06-01', '2022-09-30'),
    ('2023-06-01', '2023-09-30'),
]

OPTICAL_BANDS = ['B2', 'B3', 'B4', 'B8', 'B11', 'B12']  # matches module2_dataset.py OPTICAL_BAND_ORDER exactly
PATCH_SIZE_PX = 256
CRS = 'EPSG:32645'
MAX_PRE_GAP_SEARCH_DAYS = 90
GT_MAX_CLOUD_FRACTION = 0.05
CHUNK_SIZE = 15

brahmapur_patches = ee.FeatureCollection(f'users/{GEE_USERNAME}/brahmapur_patch_grid')
mahanadi_patches = ee.FeatureCollection(f'users/{GEE_USERNAME}/mahanadi_patch_grid')
all_patches = brahmapur_patches.merge(mahanadi_patches)

# ── helpers ────────────────────────────────────────────────────────────────────

def get_s2_with_cloud_prob(start, end, geom):
    s2_sr = (ee.ImageCollection('COPERNICUS/S2_SR_HARMONIZED')
             .filterBounds(geom)
             .filterDate(start, end)
             .select(OPTICAL_BANDS))

    s2_cloud = (ee.ImageCollection('COPERNICUS/S2_CLOUD_PROBABILITY')
                .filterBounds(geom)
                .filterDate(start, end))

    joined = ee.Join.saveFirst('cloud_prob_img').apply(
        primary=s2_sr,
        secondary=s2_cloud,
        condition=ee.Filter.equals(
            leftField='system:index',
            rightField='system:index'
        )
    )

    def add_cloud_band(obj):
        img = ee.Image(obj)  # explicit cast — Join.saveFirst returns generic objects
        cloud_img = ee.Image(img.get('cloud_prob_img')).select('probability')
        cloud_mask = cloud_img.gt(40).rename('CLOUD_MASK').toUint16()
        return img.addBands(cloud_mask)

    return ee.ImageCollection(joined.map(add_cloud_band))

    def add_cloud_band(img):
        cloud_img = ee.Image(img.get('cloud_prob_img')).select('probability')
        cloud_mask = cloud_img.gt(40).rename('CLOUD_MASK')  # threshold: >40% prob = cloudy pixel
        return img.addBands(cloud_mask)

    return ee.ImageCollection(joined.map(add_cloud_band))


def compute_patch_cloud_fraction(img, geom):
    """
    Computes mean cloud probability over the patch geometry.
    Returns a server-side number (not yet evaluated).
    """
    cloud_prob_img = ee.Image(img.get('cloud_prob_img')).select('probability')
    mean_dict = cloud_prob_img.reduceRegion(
        reducer=ee.Reducer.mean(),
        geometry=geom,
        scale=10,
        maxPixels=1e6
    )
    return mean_dict.getNumber('probability')


def get_best_reference(geom, window_start):
    """
    Finds the clearest (lowest mean cloud probability over patch) S2 scene
    in the MAX_PRE_GAP_SEARCH_DAYS before window_start.
    Returns dict with keys: found (bool), date (str), cloud_fraction (float)
    — evaluated server-side via getInfo().
    """
    end_date = window_start
    start_date = (ee.Date(window_start)
                  .advance(-MAX_PRE_GAP_SEARCH_DAYS, 'day')
                  .format('YYYY-MM-dd')
                  .getInfo())

    coll = get_s2_with_cloud_prob(start_date, end_date, geom)
    size = coll.size().getInfo()
    if size == 0:
        return {'found': False, 'date': None, 'cloud_fraction': None}

    # Sort by scene-level CLOUDY_PIXEL_PERCENTAGE first (cheap metadata sort)
    # then refine with per-patch reduceRegion on the top candidate only
    # This avoids running reduceRegion on every scene in the collection
    sorted_coll = coll.sort('CLOUDY_PIXEL_PERCENTAGE')
    best_img = ee.Image(sorted_coll.first())

    cf = compute_patch_cloud_fraction(best_img, geom).getInfo()
    if cf is None:
        return {'found': False, 'date': None, 'cloud_fraction': None}

    date_ms = best_img.get('system:time_start').getInfo()
    date_str = time.strftime('%Y-%m-%d', time.gmtime(date_ms / 1000))
    return {'found': True, 'date': date_str, 'cloud_fraction': cf}


def get_best_ground_truth(geom, window_start, window_end):
    """
    Finds the clearest S2 scene strictly inside the gap window
    with mean patch cloud fraction < GT_MAX_CLOUD_FRACTION (0.05).
    Returns dict with keys: found (bool), date (str), cloud_fraction (float).
    """
    coll = get_s2_with_cloud_prob(window_start, window_end, geom)
    size = coll.size().getInfo()
    if size == 0:
        return {'found': False, 'date': None, 'cloud_fraction': None}

    sorted_coll = coll.sort('CLOUDY_PIXEL_PERCENTAGE')

    # Iterate through candidates until one passes the per-patch threshold
    # Limit to top 5 candidates to avoid excessive getInfo() calls
    candidates = sorted_coll.toList(5).getInfo()
    for candidate in candidates:
        img = ee.Image(sorted_coll.toList(5).get(candidates.index(candidate)))
        cf = compute_patch_cloud_fraction(img, geom).getInfo()
        if cf is not None and cf < (GT_MAX_CLOUD_FRACTION * 100):  # cloud_prob is 0-100
            date_ms = candidate['properties']['system:time_start']
            date_str = time.strftime('%Y-%m-%d', time.gmtime(date_ms / 1000))
            return {'found': True, 'date': date_str, 'cloud_fraction': cf}

    return {'found': False, 'date': None, 'cloud_fraction': None}


# ── main loop ──────────────────────────────────────────────────────────────────

print('Fetching patch list from server...')
patch_list_raw = all_patches.toList(all_patches.size()).getInfo()
print(f'Patches loaded: {len(patch_list_raw)}')

ref_export_count = 0
gt_export_count = 0
ref_skipped_count = 0
gt_skipped_count = 0
failed_submissions = []

# Chunk the patch list to avoid timeout on expensive per-patch cloud queries
chunks = [patch_list_raw[i:i+CHUNK_SIZE] for i in range(0, len(patch_list_raw), CHUNK_SIZE)]
print(f'Processing {len(chunks)} chunks of up to {CHUNK_SIZE} patches each...')

for chunk_idx, chunk in enumerate(chunks):
    print(f'\nChunk {chunk_idx+1}/{len(chunks)} ({len(chunk)} patches)...')

    for patch_feature in chunk:
        patch_id = patch_feature['properties']['patch_id']
        patch_geom = ee.Geometry(patch_feature['geometry'])

        for window_start, window_end in GAP_WINDOWS:

            # ── REFERENCE export ───────────────────────────────────────────
            ref_info = get_best_reference(patch_geom, window_start)

            if not ref_info['found']:
                ref_skipped_count += 1
                print(f'  REF SKIP: {patch_id} {window_start} — no usable scene in prior 90 days')
            else:
                ref_date = ref_info['date']
                ref_coll = get_s2_with_cloud_prob(
                    (ee.Date(window_start)
                     .advance(-MAX_PRE_GAP_SEARCH_DAYS, 'day')
                     .format('YYYY-MM-dd')
                     .getInfo()),
                    window_start,
                    patch_geom
                )
                ref_img = ee.Image(ref_coll.sort('CLOUDY_PIXEL_PERCENTAGE').first())

                # Export: 6 optical bands + CLOUD_MASK (7 bands total, in order)
                # module2_dataset.py reads ref[:6] as optical, ref[6:7] as cloud mask
                ref_export_img = ref_img.select(OPTICAL_BANDS + ['CLOUD_MASK']).clip(patch_geom)
                description = f's2ref_{patch_id}_{window_start}'

                try:
                    task = ee.batch.Export.image.toDrive(
                        image=ref_export_img,
                        description=description,
                        folder='sar_optical_dataset/optical_reference',
                        fileNamePrefix=description,
                        region=patch_geom,
                        dimensions=f'{PATCH_SIZE_PX}x{PATCH_SIZE_PX}',
                        crs=CRS,
                        maxPixels=1e9
                    )
                    task.start()
                    ref_export_count += 1
                except Exception as e:
                    failed_submissions.append((description, str(e)))

            # ── GROUND TRUTH export ────────────────────────────────────────
            gt_info = get_best_ground_truth(patch_geom, window_start, window_end)

            if not gt_info['found']:
                gt_skipped_count += 1
                # Expected — most patch/windows will have no clear scene during monsoon
            else:
                gt_date = gt_info['date']
                gt_coll = get_s2_with_cloud_prob(window_start, window_end, patch_geom)
                gt_img = ee.Image(gt_coll.sort('CLOUDY_PIXEL_PERCENTAGE').first())

                gt_export_img = gt_img.select(OPTICAL_BANDS).clip(patch_geom)
                description = f's2gt_{patch_id}_{window_start}'

                try:
                    task = ee.batch.Export.image.toDrive(
                        image=gt_export_img,
                        description=description,
                        folder='sar_optical_dataset/optical_ground_truth',
                        fileNamePrefix=description,
                        region=patch_geom,
                        dimensions=f'{PATCH_SIZE_PX}x{PATCH_SIZE_PX}',
                        crs=CRS,
                        maxPixels=1e9
                    )
                    task.start()
                    gt_export_count += 1
                except Exception as e:
                    failed_submissions.append((description, str(e)))

        # Small sleep every patch to avoid hammering the API
        time.sleep(0.5)

    print(f'  Chunk {chunk_idx+1} done. Ref exports so far: {ref_export_count}, GT exports: {gt_export_count}')
    time.sleep(2)  # brief pause between chunks

# ── summary ───────────────────────────────────────────────────────────────────
print(f'\n{"="*60}')
print(f'REFERENCE exports submitted:  {ref_export_count}')
print(f'REFERENCE skipped (no scene): {ref_skipped_count}')
print(f'GROUND TRUTH exports submitted: {gt_export_count}')
print(f'GROUND TRUTH skipped (no clear scene in window): {gt_skipped_count}')
if failed_submissions:
    print(f'\nSubmission-time failures: {len(failed_submissions)}')
    for desc, err in failed_submissions[:10]:
        print(f'  {desc}: {err}')
print('\nCheck progress at https://code.earthengine.google.com/tasks')